src/unix/glegl.cpp

```diff
@@ -0,0 +1,646 @@
+                         const wxGLContext *other,
+                         const wxGLContextAttrs *ctxAttrs)
+    : m_glContext(NULL)
+{
+    const int* contextAttribs = NULL;
+
+    if ( ctxAttrs )
+    {
+        contextAttribs = ctxAttrs->GetGLAttrs();
+    }
+    else if ( win->GetGLCTXAttrs().GetGLAttrs() )
+    {
+        // If OpenGL context parameters were set at wxGLCanvas ctor, get them now
+        contextAttribs = win->GetGLCTXAttrs().GetGLAttrs();
+    }
+
+    m_isOk = false;
+
+    EGLConfig *fbc = win->GetEGLConfig();
+    wxCHECK_RET( fbc, "Invalid EGLConfig for OpenGL" );
+
+    m_glContext = eglCreateContext(wxGLCanvasEGL::GetDisplay(), fbc[0],
+                                   other ? other->m_glContext : EGL_NO_CONTEXT,
+                                   contextAttribs);
+
+    if ( !m_glContext )
+        wxLogMessage(_("Couldn't create OpenGL context"));
+    else
+        m_isOk = true;
+}
+
+wxGLContext::~wxGLContext()
+{
+    if ( !m_glContext )
+        return;
+
+    if ( m_glContext == eglGetCurrentContext() )
+        eglMakeCurrent(wxGLCanvasEGL::GetDisplay(), EGL_NO_SURFACE,
+                       EGL_NO_SURFACE, EGL_NO_CONTEXT);
+
+    eglDestroyContext(wxGLCanvasEGL::GetDisplay(), m_glContext);
+}
+
+bool wxGLContext::SetCurrent(const wxGLCanvas& win) const
+{
+    if ( !m_glContext )
+        return false;
+
+    return eglMakeCurrent(win.GetEGLDisplay(), win.GetEGLSurface(),
+                          win.GetEGLSurface(), m_glContext);
+}
+
+// ============================================================================
+// wxGLCanvasEGL implementation
+// ============================================================================
+
+// ----------------------------------------------------------------------------
+// initialization methods and dtor
+// ----------------------------------------------------------------------------
+
+wxGLCanvasEGL::wxGLCanvasEGL()
+{
+    m_config = NULL;
+    m_display = NULL;
+    m_surface = EGL_NO_SURFACE;
+    m_wl_compositor = NULL;
+    m_wl_subcompositor = NULL;
+    m_wl_frame_callback_handler = NULL;
+    m_wl_egl_window = NULL;
+    m_wl_surface = NULL;
+    m_wl_region = NULL;
+    m_wl_subsurface = NULL;
+    m_ready_to_draw = false;
+}
+
+bool wxGLCanvasEGL::InitVisual(const wxGLAttributes& dispAttrs)
+{
+    bool ret = InitConfig(dispAttrs, &m_config);
+    if ( !ret )
+    {
+        wxFAIL_MSG("Failed to get an EGLConfig for the requested attributes.");
+    }
+    return ret;
+}
+
+/* static */
+EGLDisplay wxGLCanvasEGL::GetDisplay()
+{
+    wxDisplayType type = wxDisplayNone;
+    void *display = wxGetDisplayAndType(&type);
+    EGLenum platform;
+    if (type == wxDisplayX11)
+        platform = EGL_PLATFORM_X11_EXT;
+    else if (type == wxDisplayWayland)
+        platform = EGL_PLATFORM_WAYLAND_EXT;
+    else
+        return EGL_NO_DISPLAY;
+
+    return eglGetPlatformDisplay(platform, display, NULL);
+}
+
+static void wl_global (void *data,
```

Comment:
I'd put all these function in an anonymous namespace, rather than using `static` for them, and use `wx` prefix to make it clear that these are our own functions.

Reply:
Should actually be `extern "C"`, so can't be in a namespace.

Reply:
Make them `extern "C"`, although I still don't understand `extern "C"` does, even after reading documentation.  :-)

Reply:
Basically any callbacks you pass to C code must have C linkage (i.e. be `extern "C"`). In practice, C++ linkage is the same for function pointers (but for functions, which get mangled in C++ but not in C) on all platforms I know about, but this doesn't need to be the case, in theory.

Reply:
Yeah, I get that, but I guess what I don't understand is why does it work (at least for me) without the `extern "C"`?

Reply:
The code here uses function pointers, not functions themselves, so there is no mangling involved and, as I wrote, in practice the actual calling convention for global functions is exactly the same for both C and C++, so it's not surprising that it works. In theory, however, this is undefined behaviour and as it doesn't take much to fix it, it's better to use `extern "C"` even if it's not really necessary.

Reply:
Got it, thanks!